### PR TITLE
DM-47328: Experimental command to copy files based on a graph

### PR DESCRIPTION
**Pull request by timj**
Requiers lsst/daf_butler#1111

cc/ @tgoldina I think this might do what you need. Please give it a try and let me know. I might change the name of the command etc.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

**Comment by timj** [comment]
@andy-slac it looks like I should add `QuantumGraph.get_inputs()` and `QuantumGraph.get_outputs()`  methods. Does that seem right to you? Not sure whether they even need parameters -- just return all of them, and return datastore records for inputs and update the storage classes for the outputs.

**Comment by timj** [comment]
Looking at the loops that are involved with the same loop over quanta running for both inputs and outputs, I'm edging towards a single `get_refs(self, *, include_inputs=True, include_initInputs=True, include_outputs=True, include_initOutputs=True)` approach.

File: python/lsst/pipe/base/graph/graph.py [hunk 1]
```diff
@@ -1656,3 +1658,88 @@ def init_output_run(self, butler: LimitedButler, existing: bool = True) -> None:
         self.write_configs(butler, compare_existing=existing)
         self.write_packages(butler, compare_existing=existing)
         self.write_init_outputs(butler, skip_existing=existing)
+
+    def get_refs(
```
Comment (timj) [inline python/lsst/pipe/base/graph/graph.py:1662]: @andy-slac what do you think of this? As for intermediates, I've lost track of how I find them.

Comment (andy-slac) [reply python/lsst/pipe/base/graph/graph.py:1662]: Looks good to me, if that works for you. Intermediates should be intersection of inputs and outputs?

**Comment by tgoldina** [comment]
@timj Thank you very much!

Indeed, I was able to retrieve all inputs used in a quantum graph with
```
$ pip install git+https://github.com/lsst/daf_butler@tickets/DM-47328
$ pip install git+https://github.com/lsst/pipe_base/@tickets/DM-47328
$ butler retrieve-artifacts-for-quanta --preserve-path --include-inputs --no-include-outputs tst.qgraph DATA tst_outdir
```

**Review by andy-slac — APPROVED** [review]
Looks great, I do not have an opinion on command name for CLI, `retrieve-artifacts-for-quanta` looks reasonable.

File: python/lsst/pipe/base/graph/graph.py [hunk 2]
```diff
@@ -1656,3 +1658,153 @@ def init_output_run(self, butler: LimitedButler, existing: bool = True) -> None:
...
+                internal_dataset_type = dataset_types.get(ref.datasetType.name, ref.datasetType)
+                if internal_dataset_type.storageClass_name != ref.datasetType.storageClass_name:
+                    ref = ref.overrideStorageClass(internal_dataset_type.storageClass_name)
+                return ref
+
+            # Convert output_refs to the data repository storage classes, too.
+            output_refs = {_update_ref(ref) for ref in output_refs}
+            init_output_refs = {_update_ref(ref) for ref in init_output_refs}
+
+        _LOG.info(
```
[the remark this answers is not in the document]
Comment (timj) [reply python/lsst/pipe/base/graph/graph.py:1800]: I think it's a useful message for command tools but info is too much for a general purpose API. I'll change to VERBOSE.

File: python/lsst/pipe/base/cli/cmd/commands.py [hunk 3]
```diff
@@ -92,3 +93,61 @@ def zip_from_graph(**kwargs: Any) -> None:
...
+    help="If clobber, overwrite files if they exist locally.",
+)
+@click.option(
+    "--qgraph-node-id",
+    callback=split_commas,
+    multiple=True,
+    help=unwrap(
+        """Only load a specified set of nodes when graph is
+        loaded from a file, nodes are identified by UUID
+        values. One or more comma-separated integers are
```
[the remark this answers is not in the document]
Comment (timj) [reply python/lsst/pipe/base/cli/cmd/commands.py:126]: Funnily enough that text comes from ctrl_mpexec...

[the remark this answers is not in the document]
Comment (andy-slac) [reply python/lsst/pipe/base/cli/cmd/commands.py]: Originally node IDs were integers, I guess only a part of that help string was updated when we moved to UUIDs.

File: python/lsst/pipe/base/graph/graph.py [hunk 4]
```diff
@@ -1656,3 +1658,153 @@ def init_output_run(self, butler: LimitedButler, existing: bool = True) -> None:
...
+            request_include_init_inputs = include_init_inputs
+            request_include_inputs = include_inputs
+            request_include_init_outputs = include_init_outputs
+            request_include_outputs = include_outputs
+
+        if request_include_init_inputs or request_include_init_outputs:
+            for task_def in self.iterTaskGraph():
+                if request_include_init_inputs:
+                    if in_refs := self.initInputRefs(task_def):
+                        init_input_refs.update(in_refs)
```
[the remark this answers is not in the document]
Comment (timj) [reply python/lsst/pipe/base/graph/graph.py:1727]: You mean it should be added to init_output_refs not input_refs? I see this is things like `packages`.

[the remark this answers is not in the document]
Comment (timj) [reply python/lsst/pipe/base/graph/graph.py:1727]: (but it is in the pipelines_check repo so somehow it was turning up)

[the remark this answers is not in the document]
Comment (timj) [reply python/lsst/pipe/base/graph/graph.py:1727]: Oh I see. It's a bug. I hadn't spotted that it was in the script originally when I migrated the code over to the shared version!. Will fix.

**Pull request merged**